File: api/submit.py
import json
import os
import datetime
from http.server import BaseHTTPRequestHandler
from upstash_redis import Redis

KV_KEY = "survey_responses"


def _redis():
    return Redis(
        url=os.environ["KV_REST_API_URL"],
        token=os.environ["KV_REST_API_TOKEN"],
    )


class handler(BaseHTTPRequestHandler):

    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            data = json.loads(body)

            if not isinstance(data, dict):
                self._respond(400, {"error": "Invalid payload"})
                return

            # Sanitize
            MAX_KEYS, MAX_VAL = 200, 5000
            if len(data) > MAX_KEYS:
                self._respond(400, {"error": "Too many fields"})
                return

            sanitized = {str(k)[:300]: str(v)[:MAX_VAL] for k, v in data.items() if v is not None}
            sanitized["_submitted_at"] = datetime.datetime.utcnow().isoformat() + "Z"

            # Append to Redis list
            r = _redis()
            r.rpush(KV_KEY, json.dumps(sanitized, ensure_ascii=False))
            total = r.llen(KV_KEY)

            self._respond(200, {"status": "ok", "total": total})

        except Exception as e:
            self._respond(500, {"error": str(e)})

    def do_OPTIONS(self):
        self._cors_headers()
        self.send_response(204)
        self.end_headers()

    def _respond(self, status, body):
        payload = json.dumps(body).encode()
        self._cors_headers()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)

    def _cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

File: api/submit.js
import { Redis } from '@upstash/redis';

const redis = new Redis({
  url: process.env.KV_REST_API_URL,
  token: process.env.KV_REST_API_TOKEN,
});

const KV_KEY = 'survey_responses';
const MAX_KEYS = 200;
const MAX_VAL = 5000;

export default async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return res.status(204).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const data = req.body;
    if (!data || typeof data !== 'object' || Array.isArray(data)) {
      return res.status(400).json({ error: 'Invalid payload' });
    }
    if (Object.keys(data).length > MAX_KEYS) {
      return res.status(400).json({ error: 'Too many fields' });
    }

    const sanitized = {};
    for (const [k, v] of Object.entries(data)) {
      sanitized[String(k).slice(0, 300)] = String(v ?? '').slice(0, MAX_VAL);
    }
    sanitized._submitted_at = new Date().toISOString();

    await redis.rpush(KV_KEY, JSON.stringify(sanitized));
    const total = await redis.llen(KV_KEY);

    res.setHeader('Access-Control-Allow-Origin', '*');
    return res.status(200).json({ status: 'ok', total });
  } catch (err) {
    console.error('Submit error:', err);
    return res.status(500).json({ error: err.message });
  }
}
